core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -537,9 +553,10 @@ sealed abstract class IO[+E, +A] extends Serializable { self =>
     self
       .map(f)
       .attempt
-      .raceWith[E, Either[E, B], B, B](IO.now[B](z).delay(duration))(
-        (either: Either[E, B], right: Fiber[E, B]) => right.interrupt *> either.fold(IO.fail, IO.now),
-        (b: B, left: Fiber[E, Either[E, B]]) => left.interrupt *> IO.now(b)
+      .raceWith(IO.now[B](z).delay(duration))(
```

Comment:
While you're in here, can you change this so that if `self` is terminated, the action times out as well.

It strikes me that would be easier if we had an even more powerful `raceWith` operation that would call `leftDone` and then `rightDone` in some order and then could decide to do whatever with a promise and the fibers that are passed in (or maybe there's a better way to generalize it).

Reply:
This one is trickier 🤠  If the `arbiter` function was provided by the caller, then we could implement both strategies, i.e. fail fast or fail last. For fail fast, it becomes:
```scala
def arbiterFast(f, winner, loser, race, done) = {
  case ExitResult.Completed(_) => f(winner, loser).map(done.done(_))
  case ExitResult.Failed(e, ts) => loser.interrupt *> IO.fail0(e, ts)
  case ExitResult.Terminated(ts) => loser.interrupt *> IO.terminate(ts)
}
```
It's ugly though, would require some massaging. With that:
```scala
def timeout[B](z: B)(f: A => B)(duration: Duration): IO[E, B] =
  (self.map(f) raceWith IO.now(z).delay(duration))(arbiterFast, _.join <* _.interrupt)
```